sqlmesh/core/macros.py

```diff
@@ -349,6 +354,65 @@ def engine_adapter(self) -> EngineAdapter:
             )
         return self.locals["engine_adapter"]
 
+    def _coerce(self, expr: exp.Expression, typ: t.Type) -> t.Any:
+        """Coerces the given expression to the specified type on a best-effort basis."""
+        try:
+            if typ is None:
+                return expr
+            base = t.get_origin(typ) or typ
+            if isinstance(expr, base):
+                return expr
+            if t.get_origin(typ) is t.Union:
+                for t_ in t.get_args(typ):
+                    try:
+                        return self._coerce(expr, t_)
+                    except Exception:
+                        pass
+                raise SQLMeshError(f"Failed to coerce expression '{expr}' to type '{typ}'.")
+            if issubclass(base, exp.Expression):
+                d = Dialect.get_or_raise(self.dialect)
+                into = base if base in d.parser().EXPRESSION_PARSERS else None
+                coerced = parse_one(
+                    expr.this if isinstance(expr, exp.Literal) else expr.sql(), into=into
+                )
+                if isinstance(coerced, base):
+                    return coerced
+                else:
+                    raise SQLMeshError(f"Failed to coerce expression '{expr}' to type '{typ}'.")
+            if base in (int, float, str) and isinstance(expr, exp.Literal):
+                return base(expr.this)
+            if base is bool and isinstance(expr, exp.Boolean):
+                return expr.this
+            if base is str and isinstance(expr, exp.Expression):
+                return expr.sql(self.dialect)
+            if base is tuple and isinstance(expr, (exp.Tuple, exp.Array)):
+                generic = t.get_args(typ)
+                if not generic:
+                    return tuple(expr.expressions)
```

Comment:
not every expression defines expressions

Reply:
we already asserted its an `exp.Tuple` or `exp.Array`